[Web:Discover] Bring back db service checker

kimlisa opened the pull request:
part of https://github.com/gravitational/teleport/issues/16858

Note: some files are large so you have to click `load diff` to see the file

#### Description
- Allows updating by other select database fields
- Re-enables the db service checker
- ~~Adds event emitter for failure to fetch service and update database~~
- Fixed the db service label checker to do exact matching (new db labels should == db service matcher labels)
- Placed a loose check for proper AWS endpoint connection format, as this caused db service to silently not pick up the database, despite matching labels
- Allow user to update db fields (only after they successfully created a DB, but they encountered an error later on (eg network error or db service not picking up the db), and they go back to the editing screen and wish to change things
- Split register db to two substeps as screenshotted below, this was to prevent user from changing the db name as that is not allowed with update action

#### Screenshot
First step, define db name

![image](https://user-images.githubusercontent.com/43280172/216017443-6e52f224-80c4-4416-a933-68cf5ddcf17a.png)

Second step (last), define other db fields (all these are updatable)

![image](https://user-images.githubusercontent.com/43280172/216017566-26ca533e-6d80-4617-bf85-9fee07af092e.png)

kimlisa committed:
[Backend]: allow more db fields to be updatable

kimlisa committed:
Add preventing duplicate keys for label creator

kimlisa committed:
Add sort compare fn helper

kimlisa committed:
Update db request update object

kimlisa committed:
Update CreateDatabase component

- Uncomment and touch up db service checker
- Fix db service label checking
- Allow user to make changes to updatable db fields
- Split defining db name and the other fields into
  two views, to prevent user from changing the db name
  which is not allowed with the update action

kimlisa commented:
@marcoandredinis may i get your review for this particular commit please: https://github.com/gravitational/teleport/pull/21062/commits/c0e09586fd03563c63df4e54987b95c36024f126

kimlisa committed:
Address CR: use nil value to determine CACert update

kimlisa commented on lib/web/databases.go:
```diff
@@ -141,16 +117,42 @@ func (h *Handler) handleDatabaseCreate(w http.ResponseWriter, r *http.Request, p
+	// the provided CA.
+	// When set to false, updates all other fields and
+	// skips updating and validating CA.
+	UpdateCACert bool       `json:"updateCaCert,omitempty"`
```
brilliant idea, done

marcoandredinis commented on lib/web/databases.go:
```diff
@@ -141,16 +117,40 @@ func (h *Handler) handleDatabaseCreate(w http.ResponseWriter, r *http.Request, p
 // updateDatabaseRequest contains some updatable fields of a database resource.
 type updateDatabaseRequest struct {
-	CACert string `json:"ca_cert,omitempty"`
+	// CACert when non-nil means that this request is to
+	// only update the CA and validates it.
+	// When nil, skips validating CACert and updates all other
+	// fields.
```
I wouldn't turn this into an OR (update CACert OR other fields)
If any of the other fields is present, then they should be updated

We can use the same method for the other fields: if nil, then ignore its value and update the other fields.

ryanclark commented on lib/web/databases.go:
```diff
@@ -141,16 +117,40 @@ func (h *Handler) handleDatabaseCreate(w http.ResponseWriter, r *http.Request, p
+	// only update the CA and validates it.
+	// When nil, skips validating CACert and updates all other
+	// fields.
+	CACert *string    `json:"ca_cert,omitempty"`
+	Labels []ui.Label `json:"labels,omitempty"`
+	URI    string     `json:"uri,omitempty"`
+	AWSRDS *awsRDS    `json:"awsRds,omitempty"`
```
Can the JSON fields here be consistent? Either all in snake_case or camelCase

kimlisa replied:
camelCased `ca_cert`

kimlisa committed:
Few more touch ups
- address CR: don't OR update db request field CACert
  and use consistent camelCasing
- remove the skip button for deploying service since
  we have the service checker now, which auto skips it
  for user if detects service
- refactor label matching for both register database
  and deploy service steps

ryanclark commented on lib/web/databases.go:
```diff
@@ -141,16 +117,37 @@ func (h *Handler) handleDatabaseCreate(w http.ResponseWriter, r *http.Request, p
+		if _, err := tlsutils.ParseCertificatePEM([]byte(*r.CACert)); err != nil {
+			return trace.BadParameter("could not parse provided CA as X.509 PEM certificate")
+		}
+		return nil
```
Should we return `nil` here? It would skip the checks below if the `CACert` stuff is correct

kimlisa replied:
gah i missed that thanks: https://github.com/gravitational/teleport/pull/21062/commits/73666f008542c3dce0415cf4ce2767b99871d3a7

ryanclark commented on lib/web/databases.go:
```diff
@@ -141,16 +117,37 @@ func (h *Handler) handleDatabaseCreate(w http.ResponseWriter, r *http.Request, p
+		if r.AWSRDS.AccountID == "" {
+			return trace.BadParameter("missing aws rds field account id")
+		}
+		return nil
```
Same here

marcoandredinis reviewed (approved):
LGTM

As I understand this is mostly a revert commit because the db service checker was implemented and reverted because of the backend.

The new backend bits to allow the partial update of fields LGTM

kimlisa committed:
Address CR: make all fields updatable

kimlisa committed:
Update service polling after rebase

kimlisa commented:
closing and reopening to retrigger CI
